int32 mode
int32[] data
string backStr
string rawStr